src/main.rs

```diff
@@ -271,47 +348,36 @@ fn read_cards() -> std::io::Result<Vec<Card>> {
     Ok(cards)
 }
 
-fn controllers(config: &Config, filter: bool) -> std::io::Result<Vec<CardController>> {
+fn controllers(config: &Config, filter: bool) -> std::io::Result<Vec<HwMon>> {
     Ok(read_cards()?
         .into_iter()
-        .filter(|card| {
-            !filter
-                || config
-                    .cards()
-                    .iter()
-                    .find(|name| name.0 == card.0)
-                    .is_some()
-        })
-        .map(|card| CardController::new(card).unwrap())
-        .filter(|controller| !filter || controller.hw_mon.is_amd())
-        .filter(|reader| {
-            !filter
-                || reader
-                    .hw_mon
-                    .name()
-                    .ok()
-                    .filter(|s| s.as_str() == "amdgpu")
-                    .is_some()
-        })
+        .filter(|card| !filter || config.cards().iter().find(|name| **name == *card).is_some())
+        .filter_map(|card| hw_mon::open_hw_mon(card).ok())
+        .filter(|hw_mon| !filter || { hw_mon.is_amd() })
+        .filter(|hw_mon| !filter || hw_mon.name_is_amd())
         .collect())
 }
 
 fn service(config: Config) -> std::io::Result<()> {
     let mut controllers = controllers(&config, true)?;
+    let mut cache = std::collections::HashMap::new();
     loop {
-        for controller in controllers.iter_mut() {
-            let gpu_temp = controller.hw_mon.gpu_temp().unwrap_or_default();
+        for hw_mon in controllers.iter_mut() {
+            let gpu_temp = hw_mon.max_gpu_temp().unwrap_or_default();
+            log::info!("Current {} temperature: {}", hw_mon.card, gpu_temp);
+            let last = *cache.entry(*hw_mon.card).or_insert(1_000f64);
 
-            let speed = config.speed_for_temp(gpu_temp);
-            let target_pwm = (speed as f32 * 2.55).round() as u32;
-            if controller.hw_mon.pwm_min() > target_pwm || controller.hw_mon.pwm_max() < target_pwm {
+            if ((last - 0.001f64)..(last + 0.001f64)).contains(&gpu_temp) {
+                log::info!("temperature didn't changed");
```

Comment:
```suggestion
                log::info!("Temperature didn't change");
```

Would consider changing this to `log::debug` so it doesn't spam the system log.